### PR TITLE
feat(compiler): index arrays, maps, and Json with `a[b]` syntax

### LINKED ISSUE #5702: Support array `[]` accessors
### Use Case

Instead of

```js
let arr = MutArray<str>["a", "b", "c"];
assert(arr.at(0) == "a");
arr.set(0, "A");
```

You can write

```js
let arr = ["a", "b", "c"];
assert(arr[0] == "a");
arr[0] = "A";
```

### Proposed Solution

_No response_

### Implementation Notes

_No response_

### Component

Language Design, Compiler

### Community Notes

* Please vote by adding a 👍 reaction to the issue to help us prioritize.
* If you are interested to work on this issue, please leave a comment.
* If this issue is labeled **needs-discussion**, it means the spec has not been finalized yet. Please reach out on the #dev channel in the [Wing Slack](https://t.winglang.io/slack).

### PR DESCRIPTION
**Pull request by Chriscbr**
Wing now supports a dedicated array indexing syntax `arr[index]` for getting and setting values within arrays, maps, JSON values, and strings. Here are some examples:

```js
// accessing values to perform simple lookups
let arr = MutArray<num>[3, 5];
assert(arr[0] == 3);
assert(arr[1] == 5);
assert(arr[-1] == 5);
assert(arr[-2] == 3);

arr[42]; // throws an error

// setting values
arr[0] = 42; 
arr[1] += 3.5;
```

Here are the types that can be indexed:

* `Array` and `MutArray` - accepts a `num` index
* `Map` and `MutMap` - accepts a `str` index
* `str` - accepts a `num` index
* `Json` and `MutJson` - accepts `num` and `str` index values

When accesses are performed, an error is thrown if the index is out of bounds or the key is invalid. If you want to obtain a value defensively (e.g. you want to see if a `Json` has a particular key or not, without a runtime exception), you can fall back to dedicated methods like `array.tryAt(0)` or `myjson.tryGet("key")` which return optional values.

Closes #5702

## Checklist

- [x] Title matches [Winglang's style guide](https://www.winglang.io/contributing/start-here/pull_requests#how-are-pull-request-titles-formatted)
- [x] Description explains motivation and solution
- [x] Tests added (always)
- [ ] Docs updated (only required for features)
- [ ] Added `pr/e2e-full` label if this feature requires end-to-end testing

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Wing Cloud Contribution License](https://github.com/winglang/wing/blob/main/CONTRIBUTION_LICENSE.md)*.

### CODE REVIEW

File: libs/wingsdk/src/helpers.ts [hunk 1]
```diff
@@ -55,6 +55,100 @@ export function unwrap<T>(value: T): T | never {
...
+
+  return obj[index];
+}
+
+export function assign(
+  obj: any,
+  index: string | number,
+  kind: "=" | "+=" | "-=",
+  value: any
+) {
```
Comment (MarkMcCulloh) [inline libs/wingsdk/src/helpers.ts:104]: nit: there's some duplication between `assign` and `lookup`

File: examples/tests/valid/indexing.test.w [hunk 2]
```diff
@@ -0,0 +1,68 @@
+bring cloud;
+bring expect;
+
+let arr = Array<num>[1, 2, 3];
+
+assert(arr[0] == 1);
```
Comment (MarkMcCulloh) [inline examples/tests/valid/indexing.test.w:6]: Spec update? Doc update would be nice as well, but IMO probably fine to get this PR in first.

**Review by MarkMcCulloh — APPROVED** [review]
Noice

File: examples/tests/valid/indexing.test.w
Comment (yoav-steinberg) [inline examples/tests/valid/indexing.test.w:1]: 1. Maybe add tests with an expression as the index: `arr[1+2]` or `my_map[some_str.substring(1,1)]`.
2. Maybe some tests regarding indexing into an indexed value (to verify precedence works correctly): `arr[x][y][z]`.

File: libs/wingc/src/ast.rs [hunk 3]
```diff
@@ -742,6 +742,8 @@ pub enum Reference {
 		property: Symbol,
 		optional_accessor: bool,
 	},
+	/// A reference to an accessed member of an object `expression[x]`
+	ElementAccess { object: Box<Expr>, index: Box<Expr> },
```
Comment (yoav-steinberg) [inline libs/wingc/src/ast.rs:746]: I'd like to re-thing this as a "reference" and perhaps make `ElementAccess` just another `ExprKind`. References are are expressions that have a name but `ElementAccess` is just two expressions (object and index) with no name: `[1,2,3,4][0]` should evaluate to 1.

File: libs/wingc/src/type_check.rs [hunk 4]
```diff
@@ -4485,30 +4493,112 @@ impl<'a> TypeChecker<'a> {
...
+				| Type::Optional(_)
+				| Type::Set(_)
+				| Type::MutSet(_)
+				| Type::Function(_)
+				| Type::Class(_)
+				| Type::Interface(_)
+				| Type::Struct(_)
+				| Type::Enum(_) => self.spanned_error(
+					variable,
+					format!("Unsupported reassignment of element of type {}", var_type),
```
Comment (yoav-steinberg) [inline libs/wingc/src/type_check.rs:4571]: Was there a test for this?

File: libs/wingc/src/type_check.rs [hunk 5]
```diff
@@ -5793,9 +5900,97 @@ impl<'a> TypeChecker<'a> {
...
+					Type::Anything => (
+						ResolveReferenceResult::Location(self.types.anything(), self.types.anything()),
+						instance_phase,
+					),
+					Type::Unresolved => (
+						ResolveReferenceResult::Location(self.types.error(), self.types.error()),
+						instance_phase,
+					),
+					Type::Inferred(_) => (
+						ResolveReferenceResult::Location(self.types.error(), self.types.error()),
```
Comment (yoav-steinberg) [inline libs/wingc/src/type_check.rs:5968]: Is this really an error? Not sure how this works exactly.

**Review by yoav-steinberg — COMMENTED** [review]
:sparkles:

**Review by eladb — APPROVED** [review]
hell yeah!

File: libs/wingc/src/ast.rs (hunk 3 above)
Comment (yoav-steinberg) [reply libs/wingc/src/ast.rs:746]: The difference being that anything after the `.` is a name. Json is the exception with it's dotted access to non-fields, but for any typed "object" access you always access a name. Even in json there's no real way to access an expression after the `.`: `my_json.this_cant_be_an_expression`.

Comment (MarkMcCulloh) [reply libs/wingc/src/ast.rs:746]: > The difference being that anything after the . is a name.

True, but that's more of an optimization, both for our implementation and for readability. We could easily allow `a."b"` or even `a."{some string expression}"`. Not saying we should, but we could. 
The crux of my concern is that I'm not sure we're actually making use of idea that "References are expressions that have a name" anyways.

Again I'm actually in favor of breaking down `reference` into more useful parts. Like considering lvalues and rvalues for expressions and/or seeing `.` and `[]` as operators.

File: libs/wingc/src/type_check.rs (hunk 4 above)
Comment (Chriscbr) [reply libs/wingc/src/type_check.rs:4571]: I didn't, but I can try to add one. It's actually kinda tricky coming up with examples that produce this error, but it seems to correspond to degenerate cases like:

```
let var x = 3 + "hello";
x["hi"] = 4; // error: Unsupported reassignment of element of type unresolved
```

File: libs/wingc/src/type_check.rs (hunk 5 above)
Comment (Chriscbr) [reply libs/wingc/src/type_check.rs]: I wasn't sure how to get inferred types working here so for now I've made it an error (I've added a test case just now). Open to improving this

File: libs/wingsdk/src/helpers.ts [hunk 6]
```diff
@@ -55,6 +55,100 @@ export function unwrap<T>(value: T): T | never {
...
+  if (typeof index !== "string" && typeof index !== "number") {
+    throw new TypeError(
+      `Index must be a string or number (found "${typeof index}")`
+    );
+  }
+
+  if (typeof index === "number") {
+    if (
+      !Array.isArray(obj) &&
+      !Buffer.isBuffer(obj) &&
```
[the remark this answers is not in the document]
Comment (Chriscbr) [reply libs/wingsdk/src/helpers.ts:68]: Ah, I think copilot might generated this lol

File: libs/wingsdk/src/helpers.ts (hunk 1 above)
Comment (Chriscbr) [reply libs/wingsdk/src/helpers.ts:84]: Refactored a little bit

File: examples/tests/valid/indexing.test.w (hunk 2 above)
Comment (Chriscbr) [reply examples/tests/valid/indexing.test.w:6]: 👍 Added a section to the language reference

File: examples/tests/valid/indexing.test.w
Comment (Chriscbr) [reply examples/tests/valid/indexing.test.w:1]: Added

File: libs/wingc/src/ast.rs (hunk 3 above)
Comment (yoav-steinberg) [reply libs/wingc/src/ast.rs:748]: Right, I wasn't thinking of the left side of the assignment. I recall explaining to someone in the past that in our AST "reference" is actually "lvalue", so I'm convinced what you did is the best option given the current design.
Maybe we should change `Reference` to `LValue`..

**Pull request merged**